### PR TITLE
Add reStructedText support

### PR DESCRIPTION
**Pull request by lunny**
should be part of #374, see below screenshots.

### CODE REVIEW

**Comment by lunny** [comment]
Tests added and rebased.

**Comment by lafriks** [comment]
LGTM

**Comment by JonasFranzDEV** [comment]
@lafriks LGTM seems not to work

File: modules/markup/reStructuredText/reStructuredText.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+func (Parser) Name() string {
+	return "reStructuredText"
+}
+
+// Extensions implements markup.Parser
+func (Parser) Extensions() []string {
+	return []string{".rst"}
+}
+
+// Render renders reStructuredText rawbytes to HTML
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText.go]: Is not `rawbytes` in comment redudant? And should be unified with `RenderString` comment.

File: modules/markup/reStructuredText/reStructuredText.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+func init() {
+	markup.RegisterParser(Parser{})
+}
+
+// Parser implements markup.Parser for reStructuredText
+type Parser struct {
+}
+
+// Name implements markup.Parser
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText.go]: I think function comment should tell what that function is doing, not which interface it fulfils.

File: modules/markup/reStructuredText/reStructuredText.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+func (Parser) Name() string {
+	return "reStructuredText"
+}
+
+// Extensions implements markup.Parser
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText.go]: Same as for `Name()`

File: modules/markup/reStructuredText/reStructuredText_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+	// TODO: gorst didn't support relative link.
+	/*lnk := markup.URLJoin(AppSubURL, "WikiPage")
+	test("WikiPage_\n\n.. _WikiPage: WikiPage\n",
+		`<p><a href="`+lnk+`">WikiPage</a></p>`)*/
+}
+
+func TestRender_Images(t *testing.T) {
+	setting.AppURL = AppURL
+	setting.AppSubURL = AppSubURL
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText_test.go]: Same as above.

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Comment by lunny** [comment]
@Morlinest all done.

File: modules/markup/reStructuredText/reStructuredText.go [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+	return []string{".rst"}
+}
+
+// Render renders reStructuredText bytes to HTML
+func Render(rawBytes []byte, urlPrefix string, metas map[string]string, isWiki bool) []byte {
+	p := gorst.NewParser(nil)
+	var b bytes.Buffer
+	w := bufio.NewWriter(&b)
+	p.ReStructuredText(bytes.NewReader(rawBytes), gorst.ToHTML(w))
+	w.Flush()
```
Comment (ethantkoenig) [inline modules/markup/reStructuredText/reStructuredText.go]: `w.Flush()` returns an error

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
IMO, using a third-party library that is "experimental/highly under development" ([README](https://github.com/hhatto/gorst#gorst)) is not a good idea. Would it be worth finding a more stable library, or waiting until this one becomes more mature?

**Comment by lunny** [comment]
@ethantkoenig Yes, but this is the most mature library I can find. Also, if #2570 could be merged, external tool could override internal library.

File: modules/markup/reStructuredText/reStructuredText.go (hunk 5 above)
Comment (lunny) [reply modules/markup/reStructuredText/reStructuredText.go]: done.

File: modules/markup/reStructuredText/reStructuredText_test.go [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+func TestRender_StandardLinks(t *testing.T) {
+	var appURL = setting.AppURL
+	var appSubURL = setting.AppSubURL
+	setting.AppURL = AppURL
+	setting.AppSubURL = AppSubURL
+	defer func() {
+		setting.AppURL = appURL
+		setting.AppSubURL = appSubURL
+	}()
```
[the remark this answers is not in the document]
Comment (Morlinest) [reply modules/markup/reStructuredText/reStructuredText_test.go:33]: Now, when I look at whole change at once... Why you need to change settings? I see only `setting.AppSubURL` is used in `test()`.

File: modules/markup/reStructuredText/reStructuredText_test.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+	"github.com/stretchr/testify/assert"
+)
+
+const AppURL = "http://localhost:3000/"
+const Repo = "go-gitea/gitea"
+const AppSubURL = AppURL + Repo + "/"
+
+func test(t *testing.T, input, expected string) {
+	buffer := RenderString(input, setting.AppSubURL, nil, false)
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText_test.go:21]: Use `setting.AppSubURL` as another `test()` argument. E.g. `test(t *testing.T, input, expected, appSubURL string)`

File: modules/markup/reStructuredText/reStructuredText_test.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	"testing"
+
+	"code.gitea.io/gitea/modules/setting"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const AppURL = "http://localhost:3000/"
+const Repo = "go-gitea/gitea"
+const AppSubURL = AppURL + Repo + "/"
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText_test.go:18]: If you don't need all constants (look at comment for `defer`), merge them together. Also I think you can use unexported version. E.g. `const appSubURL = "http://localhost:3000/go-gitea/gitea/"`.

File: modules/markup/reStructuredText/reStructuredText_test.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+	googleRendered := `<p><a href="https://google.com/">reStructuredText</a></p>`
+	test(t, "reStructuredText_\n\n.. _reStructuredText: https://google.com/\n", googleRendered)
+
+	// TODO: gorst didn't support relative link.
+	/*lnk := markup.URLJoin(AppSubURL, "WikiPage")
+	test("WikiPage_\n\n.. _WikiPage: WikiPage\n",
+		`<p><a href="`+lnk+`">WikiPage</a></p>`)*/
+}
+
+func TestRender_Images(t *testing.T) {
```
Comment (Morlinest) [inline modules/markup/reStructuredText/reStructuredText_test.go:44]: If you don' need to change settings (look at `defer` comment above), you can merge both tests together, remove `setting` variable changes and `defer` and use `t.Run()` (just create array of test cases) to run tests in parallel.

**Review by Morlinest — CHANGES_REQUESTED** [review]

**Review by ethantkoenig — APPROVED** [review]
I still do not think that using an unstable library is a good idea, but I'll approve so that I don't block the PR.

**Comment by lunny** [comment]
@ethantkoenig I don't mind to put this PR later. If #2570 could be merged, it also resolved #374.

**Comment by lunny** [comment]
And maybe https://github.com/demizer/go-rst is a better choice.

**Comment by ghost** [comment]
@lunny go-rst is marked as `ABANDONED`. I think with #374 closed this could also be closed.

**Pull request closed without merging**